relations/src/shielder/deposit_and_merge.rs

```diff
@@ -0,0 +1,399 @@
+            &old_trapdoor,
+            &old_nullifier,
+            &old_note,
+        )?;
+
+        //------------------------------
+        // Check the new note arguments.
+        //------------------------------
+        let new_note = FpVar::new_input(ns!(cs, "new note"), || {
+            self.new_note.ok_or(AssignmentMissing)
+        })?;
+        let new_token_amount = FpVar::new_witness(ns!(cs, "new token amount"), || {
+            self.new_token_amount.ok_or(AssignmentMissing)
+        })?;
+        let new_trapdoor = FpVar::new_witness(ns!(cs, "new trapdoor"), || {
+            self.new_trapdoor.ok_or(AssignmentMissing)
+        })?;
+        let new_nullifier = FpVar::new_witness(ns!(cs, "new nullifier"), || {
+            self.new_nullifier.ok_or(AssignmentMissing)
+        })?;
+
+        check_note(
+            &token_id,
+            &new_token_amount,
+            &new_trapdoor,
+            &new_nullifier,
+            &new_note,
+        )?;
+
+        //----------------------------------
+        // Check the token values soundness.
+        //----------------------------------
+        let token_amount = FpVar::new_input(ns!(cs, "token amount"), || {
+            self.token_amount.ok_or(AssignmentMissing)
+        })?;
+        // some range checks for overflows?
+        let token_sum = token_amount.add(old_token_amount);
+        token_sum.enforce_equal(&new_token_amount)?;
+
+        //------------------------
+        // Check the merkle proof.
+        //------------------------
+        let merkle_root = FpVar::new_input(ns!(cs, "merkle root"), || {
+            self.merkle_root.ok_or(AssignmentMissing)
+        })?;
+        let mut leaf_index = FpVar::new_witness(ns!(cs, "leaf index"), || {
+            self.leaf_index.ok_or(AssignmentMissing)
+        })?;
+
+        let mut current_hash_bytes = old_note.to_bytes()?;
+        let mut hash_bytes = vec![current_hash_bytes.clone()];
+        let path = self.merkle_path.unwrap_or_default();
+
+        if path.len() > self.max_path_len as usize {
+            return Err(UnconstrainedVariable);
+        }
+
+        let zero = CircuitField::zero();
+
+        for i in 0..self.max_path_len {
+            let sibling = FpVar::new_witness(ns!(cs, "merkle path node"), || {
+                Ok(path.get(i as usize).unwrap_or(&zero))
+            })?;
+            let bytes: Vec<ByteVar> = if leaf_index.value().unwrap_or_default().0.is_even() {
+                [current_hash_bytes.clone(), sibling.to_bytes()?].concat()
+            } else {
+                [sibling.to_bytes()?, current_hash_bytes.clone()].concat()
+            };
+
+            current_hash_bytes = tangle_in_field::<2>(bytes)?;
+            hash_bytes.push(current_hash_bytes.clone());
+
+            leaf_index = FpVar::constant(
+                leaf_index
+                    .value()
+                    .unwrap_or_default()
+                    .div(CircuitField::from(2)),
+            );
+        }
```

Comment:
this can be extracted into a common place (to avoid duplication in `withdraw`)

Reply:
You mean `CircuitField::from(2)`?

Reply:
the whole merkle proof checking